File: 0002_add-two-numbers.py
from typing import Optional, List


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def addTwoNumbers(
        self, l1: Optional[ListNode], l2: Optional[ListNode]
    ) -> Optional[ListNode]:
        out_head = ListNode()
        carry = 0
        curr = out_head
        while (l1 != None) or (l2 != None) or (carry != 0):
            _sum = (l1.val if l1 else 0) + (l2.val if l2 else 0) + carry
            carry = _sum // 10
            curr.next = ListNode(_sum % 10)
            curr = curr.next

            if l1 is not None:
                l1 = l1.next
            if l2 is not None:
                l2 = l2.next

        return out_head.next

File: 0002_add-two-numbers.cpp
#include <memory>

// // struct ListNode {
// //   int val;
// //   ListNode *next;
// //   ListNode() : val(0), next(nullptr) {}
// //   ListNode(int x) : val(x), next(nullptr) {}
// //   ListNode(int x, ListNode *next) : val(x), next(next) {}
// };
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
  ListNode *addTwoNumbers(ListNode *l1, ListNode *l2) {
    auto head = std::make_unique<ListNode>(0);
    ListNode *curr = head.get();
    ListNode *prev = curr;

    int carry = 0;
    while (l1 != nullptr || l2 != nullptr || carry != 0) {
      const int sum = carry + (l1 ? l1->val : 0) + (l2 ? l2->val : 0);
      carry = sum / 10;

      curr->next = new ListNode(sum % 10);
      curr = curr->next;

      if (l1 != nullptr) {
        l1 = l1->next;
      }
      if (l2 != nullptr) {
        l2 = l2->next;
      }
    }

    return head->next;
  }
};
